### PR TITLE
SQL Schema update

### PR DESCRIPTION
**Pull request by LisheinLWM**
Hello,
The SQL schema as been updated such that two schemas are created within the database, one for the 24hr hour data of the previous data, and another which will contain all the data.
The data insertion into the service_type table has also been moved down to the bottom of the schema so that any data insertion occurs only after all tables are created.

### CODE REVIEW

**Review by HarryEP — APPROVED** [review]
Good, we may need to change it to account for both schemas as this only runs one schema. Even if this is duplicate code we could do yesterday_schema.sql and all_time_schema.sql?

File: pipeline/load.py [hunk 1]
```diff
@@ -1,28 +1,20 @@
...
+
 import os
+
 import pandas as pd
 import psycopg2
 from psycopg2.extras import RealDictCursor
 from psycopg2.extensions import connection
 from dotenv import load_dotenv
 
-CODES_CSV = "cancel_codes.csv"
```
Comment (Annie-EXE) [inline pipeline/load.py:9]: good use of global variables

File: pipeline/load.py [hunk 2]
```diff
@@ -1,28 +1,20 @@
...
-
-
-def write_cancel_codes(conn: connection, codes_df: pd.DataFrame):
-    '''writes the cancel codes from the dataframe to the database'''
-    records = codes_df.to_records(index=False)
-    with conn.cursor() as cur:
-        sql_query = '''
-            INSERT INTO cancel_code (code, reason, abbreviation)
-            VALUES (%s, %s, %s)'''
-        cur.executemany(sql_query, records)
```
Comment (Annie-EXE) [inline pipeline/load.py:19]: nice use of `.executemany`

File: pipeline/load.py [hunk 3]
```diff
@@ -34,10 +26,65 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
+    codes = [code for code in crs_data.flatten() if not pd.isna(code)]
+    name_data = data[['origin_stn_name', 'planned_final_destination',
+                      'destination_reached_name', 'cancellation_station_name']].values
+    names = [name for name in name_data.flatten() if not pd.isna(name)]
+    data_to_insert = list(zip(codes, names))
+
+    with conn.cursor() as cur:
+        cur.executemany("""INSERT INTO station (crs, station_name) VALUES
+                        (%s, %s) ON CONFLICT DO NOTHING;""", data_to_insert)
+    conn.commit()
```
Comment (Annie-EXE) [inline pipeline/load.py:75]: clear and readable code with consistent use of type hints

File: pipeline/load.py [hunk 4]
```diff
@@ -34,10 +26,65 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
 
     load_dotenv()
     conn = get_connection(os.environ["DB_HOST"], os.environ["DB_NAME"],
                           os.environ["DB_PASS"], os.environ["DB_USER"])
-    cancel_codes_df = pd.read_csv(CODES_CSV)
-    write_cancel_codes(conn, cancel_codes_df)
+
+    switch_between_schemas("previous_day_data")
+    # cancel_codes_df = pd.read_csv(CODES_CSV)
+    # write_cancel_codes(conn, cancel_codes_df)
```
Comment (Annie-EXE) [inline pipeline/load.py:86]: commented out code >:(

File: pipeline/load.py [hunk 5]
```diff
@@ -1,28 +1,21 @@
...
 import psycopg2
-from psycopg2.extras import RealDictCursor
+from psycopg2.extras import RealDictCursor, execute_values
 from psycopg2.extensions import connection
 from dotenv import load_dotenv
 
-CODES_CSV = "cancel_codes.csv"
-
-
-def write_cancel_codes(conn: connection, codes_df: pd.DataFrame):
```
Comment (NgawangT) [inline pipeline/load.py:12]: would insert_cancel_codes or populate_cancel_codes be a better name for the function?

File: pipeline/load.py [hunk 6]
```diff
@@ -1,28 +1,21 @@
...
-        cur.executemany(sql_query, records)
 
-    conn.commit()
+CODES_CSV = "cancel_codes.csv"
 
 
 def get_connection(host: str, db_name: str, password: str, user: str):
-    '''Connects to the database'''
+    """Connects to the database"""
+
```
Comment (NgawangT) [inline pipeline/load.py:18]: a better try except could be something like this using psycopg2 error helps catch more specific errors: def get_db_connection():
    """Establishes a connection with the PostgreSQL database."""
    try:
        conn = connect(
            user=environ.get("DATABASE_USERNAME"),
            password=environ.get("DATABASE_PASSWORD"),
            host=environ.get("DATABASE_IP"),
            port=environ.get("DATABASE_PORT"),
            database=environ.get("DATABASE_NAME"),)
        print("Database connection established successfully.")
        return conn
    except Error as err:
        print("Error connecting to database: ", err)
        sys.exit()

File: pipeline/load.py [hunk 7]
```diff
@@ -34,10 +27,68 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
+
+def write_cancel_codes(conn: connection, codes_df: pd.DataFrame):
+    """writes the cancel codes from the dataframe to the database"""
+
+    records = codes_df.to_records(index=False)
+    with conn.cursor() as cur:
+        sql_query = '''
+            INSERT INTO cancel_code (code, reason, abbreviation)
+            VALUES (%s, %s, %s) ON CONFLICT DO NOTHING;'''
```
Comment (NgawangT) [inline pipeline/load.py:45]: Nice use of On Conflict do nothing with Unique constraints to prevent duplicate data.

File: pipeline/load.py [hunk 8]
```diff
@@ -34,10 +27,68 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
 
+def switch_between_schemas(schema_name: str) -> None:
+    """Switches to the schema by the schema name provided"""
+
+    with conn.cursor() as cur:
+        cur.execute("SET search_path TO %s", (schema_name,))
+    conn.commit()
+
+
+def write_cancel_codes(conn: connection, codes_df: pd.DataFrame):
```
Comment (NgawangT) [inline pipeline/load.py:38]: would insert_cancel_codes or populate_cancel_codes be a better name for the function?

File: pipeline/load.py [hunk 9]
```diff
@@ -34,10 +27,68 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
+
+
+def insert_station_data(conn: connection, data: pd.DataFrame) -> None:
+    """Inserts station data to the database"""
+
+    col_sets = [['origin_crs', 'origin_stn_name'],
+                ['planned_final_crs', 'planned_final_destination'],
+                ['destination_reached_crs', 'destination_reached_name'],
+                ['cancellation_station_crs', 'cancellation_station_name']]
+
```
Comment (NgawangT) [inline pipeline/load.py:69]: brilliant use of sets and numpy.

File: pipeline/load.py [hunk 10]
```diff
@@ -34,10 +27,68 @@ def get_connection(host: str, db_name: str, password: str, user: str):
...
                           os.environ["DB_PASS"], os.environ["DB_USER"])
-    cancel_codes_df = pd.read_csv(CODES_CSV)
-    write_cancel_codes(conn, cancel_codes_df)
+
+    switch_between_schemas("previous_day_data")
+    # cancel_codes_df = pd.read_csv(CODES_CSV)
+    # write_cancel_codes(conn, cancel_codes_df)
+
+    data = pd.read_csv("transformed_service_data.csv")
+    # insert_company_data(conn, data)
```
Comment (NgawangT) [inline pipeline/load.py:93]: here too commented out code.

File: pipeline/transform.py [hunk 11]
```diff
@@ -127,23 +128,28 @@ def determine_if_cancel_code_is_valid(service_df: DataFrame, valid_codes_list: l
...
+    service_df = check_values_in_column_have_three_characters(
+        service_df, "destination_reached_crs", True)
+    service_df = check_values_in_column_have_three_characters(
+        service_df, "cancellation_station_crs", False)
 
     cancel_codes_url = "https://wiki.openraildata.com/index.php/Delay_Attribution_Guide"
     valid_cancel_codes = generate_list_of_valid_cancel_codes(cancel_codes_url)
-    service_df = determine_if_cancel_code_is_valid(service_df, valid_cancel_codes)
+    service_df = determine_if_cancel_code_is_valid(
+        service_df, valid_cancel_codes)
```
Comment (NgawangT) [inline pipeline/transform.py:153]: Would having output_csv_path as constant in capital and on top of the page make it more easier for other developers to find and change if need be?

**Review by NgawangT — APPROVED** [review]

**Review by HarryEP — APPROVED** [review]

**Pull request merged**